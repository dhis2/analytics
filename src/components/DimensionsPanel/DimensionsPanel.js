import React, { Component } from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'

import Filter from '../Filter/Filter'
import DimensionList from './List/DimensionList'
import { styles } from './styles/DimensionsPanel.style'

export class DimensionsPanel extends Component {
    state = { filterText: '' }

    onClearFilter = () => {
        this.setState({ filterText: '' })
    }

    onFilterTextChange = filterText => {
        this.setState({ filterText })
    }

    render() {
        const {
            dimensions,
            selectedIds,
            disabledDimension,
            lockedDimension,
            recommendedDimension,
            onDimensionClick,
            onDimensionOptionsClick,
            onDimensionDragStart,
            style,
        } = this.props

        return (
            <div style={{ ...styles.divContainer, ...style }}>
                <Filter
                    style={styles.textField}
                    placeholder={i18n.t('Filter dimensions')}
                    text={this.state.filterText}
                    onChange={this.onFilterTextChange}
                    onClear={this.onClearFilter}
                    type="search"
                />
                <DimensionList
                    dimensions={dimensions}
                    selectedIds={selectedIds}
                    filterText={this.state.filterText}
                    disabledDimension={disabledDimension}
                    lockedDimension={lockedDimension}
                    recommendedDimension={recommendedDimension}
                    onDimensionOptionsClick={onDimensionOptionsClick}
                    onDimensionClick={onDimensionClick}
                    onDimensionDragStart={onDimensionDragStart}
                />
            </div>
        )
    }
}

DimensionsPanel.propTypes = {
    dimensions: PropTypes.array.isRequired,
    disabledDimension: PropTypes.func,
    lockedDimension: PropTypes.func,
    recommendedDimension: PropTypes.func,
    selectedIds: PropTypes.array,
    style: PropTypes.object,
    onDimensionClick: PropTypes.func,
    onDimensionDragStart: PropTypes.func,
    onDimensionOptionsClick: PropTypes.func,
}

DimensionsPanel.defaultProps = {
    selectedIds: [],
    style: {},
    onDimensionClick: Function.prototype,
}

export default DimensionsPanel
