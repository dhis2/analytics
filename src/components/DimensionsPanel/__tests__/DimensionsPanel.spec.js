import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import DimensionsPanel from '../DimensionsPanel.js'

const props = {
    dimensions: [
        { id: 'id1', name: 'Dimension 1' },
        { id: 'id2', name: 'Dimension 2' },
    ],
}

test('DimensionsPanel matches the snapshot', () => {
    const { container } = render(<DimensionsPanel {...props} />)

    expect(container).toMatchSnapshot()
})

test('DimensionsPanel renders a DimensionList component', () => {
    render(<DimensionsPanel {...props} />)

    const dimensionListComponent = screen.getByTestId('dimension-list')
    expect(dimensionListComponent).toBeInTheDocument()

    const liElements = screen.getAllByTestId('dimension-item')

    expect(liElements).toHaveLength(2)

    const dim1Element = screen.getByText(props.dimensions[0].name)
    expect(dim1Element).toBeInTheDocument()

    const dim2Element = screen.getByText(props.dimensions[1].name)
    expect(dim2Element).toBeInTheDocument()
})

test('DimensionsPanel can filter the dimension list', async () => {
    const user = userEvent.setup()

    render(<DimensionsPanel {...props} />)

    const filterComponent = screen.getByPlaceholderText('Filter dimensions')
    expect(filterComponent).toBeInTheDocument()

    await user.click(filterComponent)
    await user.keyboard('1')

    const liElements = screen.getAllByTestId('dimension-item')

    expect(liElements).toHaveLength(1)

    const dim1Element = screen.getByText(props.dimensions[0].name)
    expect(dim1Element).toBeInTheDocument()

    const dim2Element = screen.queryByText(props.dimensions[1].name)
    expect(dim2Element).not.toBeInTheDocument()
})
