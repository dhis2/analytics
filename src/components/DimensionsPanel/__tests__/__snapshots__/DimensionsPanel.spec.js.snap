// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DimensionsPanel matches the snapshot 1`] = `
<div>
  <div
    style="height: 100%; display: flex; flex-direction: column; background-color: rgb(255, 255, 255); padding: 8px; overflow: hidden;"
  >
    <div
      class="container"
    >
      <div
        class="jsx-1141087211 "
        data-test="dhis2-uiwidgets-inputfield"
      >
        <div
          class="jsx-3375842594 "
          data-test="dhis2-uiwidgets-inputfield-content"
        >
          <div
            class="jsx-1220078726 "
            data-test="dhis2-uicore-box"
          >
            <div
              class="jsx-4253951613 jsx-532002680 input"
              data-test="dhis2-uicore-input"
            >
              <input
                class="jsx-4253951613 jsx-532002680 dense"
                placeholder="Filter dimensions"
                type="search"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <style />
    </div>
    <div
      class="container"
      data-test="dimension-list"
    >
      <div
        class="wrapper"
      >
        <div
          class="section"
        >
          <h3
            class="header"
          >
            Your dimensions
          </h3>
          <ul
            class="list"
          >
            <li
              class="item"
              data-test="dimension-item"
              name="Dimension 1"
            >
              <div
                class="label"
                data-test="dimension-item-button-id1"
                tabindex="0"
              >
                <div
                  class="iconWrapper"
                >
                  <svg
                    fill="none"
                    height="16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="m4.832 5.674 2.831-1.11a.913.913 0 0 1 .673 0l2.83 1.11a.548.548 0 0 1 .334.517V9.81a.548.548 0 0 1-.332.517l-2.831 1.11a.92.92 0 0 1-.673 0l-2.83-1.11a.548.548 0 0 1-.334-.517V6.19a.548.548 0 0 1 .332-.517Z"
                      stroke="#212934"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                    <path
                      d="M8 7.228 4.822 5.982M8 7.228l3.178-1.246M8 11.281V7.228"
                      stroke="#212934"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </div>
                <div
                  class="labelWrapper"
                >
                  <span
                    class="labelText"
                  >
                    <span
                      data-dimensionid="id1"
                    >
                      Dimension 1
                    </span>
                  </span>
                </div>
              </div>
            </li>
            <style />
            <li
              class="item"
              data-test="dimension-item"
              name="Dimension 2"
            >
              <div
                class="label"
                data-test="dimension-item-button-id2"
                tabindex="0"
              >
                <div
                  class="iconWrapper"
                >
                  <svg
                    fill="none"
                    height="16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="m4.832 5.674 2.831-1.11a.913.913 0 0 1 .673 0l2.83 1.11a.548.548 0 0 1 .334.517V9.81a.548.548 0 0 1-.332.517l-2.831 1.11a.92.92 0 0 1-.673 0l-2.83-1.11a.548.548 0 0 1-.334-.517V6.19a.548.548 0 0 1 .332-.517Z"
                      stroke="#212934"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                    <path
                      d="M8 7.228 4.822 5.982M8 7.228l3.178-1.246M8 11.281V7.228"
                      stroke="#212934"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </div>
                <div
                  class="labelWrapper"
                >
                  <span
                    class="labelText"
                  >
                    <span
                      data-dimensionid="id2"
                    >
                      Dimension 2
                    </span>
                  </span>
                </div>
              </div>
            </li>
            <style />
          </ul>
        </div>
      </div>
      <style />
    </div>
  </div>
</div>
`;
